### PR TITLE
Vehicle/Object Config Protection

### LINKED ISSUE #451: [Feature Request] Client support for Prevention of Vehicle Config Cloning and Configuration Saving
**Is your feature request related to a problem? Please describe.**
Some players do not want other players to be able to clone and / or save their vehicle configurations. This is reasonable, and support for it should be possible.

**Describe the solution you'd like**
In BeamMP's Multiplayer Options, a pair of options for:
* Preventing Cloning, and
* Preventing Configuration Saving

If a player has either of these enabled, other players on the server should not be able to perform the corresponding action to that player's vehicle.

**Describe alternatives you've considered**
I've personally created a client mod that can prevent anyone from cloning or saving configurations, but this includes the player themself.

**Additional context**
Making other clients aware that a certain client prevents cloning or saving of the vehicle in question may be tricky, or involve server work to support it. This could potentially be a public resource of a server plugin / client mod pairing, but building this into BeamMP itself would be much more robust.

### PR DESCRIPTION
**Pull request by Starystars67**
The aim of this PR is to add code to make the saving and cloning of others vehicles harder. 

A user can enable the setting in the options to protect their configs.

From that point all of their vehicles spawned will be labelled as protected and then another user tries to clone or save it an error will be presented and the request denied.

Needs online testing with other players.

### CODE REVIEW

**Comment by StanleyDudek** [comment]
As of https://github.com/BeamMP/BeamMP/pull/545/commits/325c861664d41980ef56497ba3fd28bc1010b871 this appears to be working for the clone current button, I tested two independent clients on two forum logins on the same official server to copy each

image

Same message presented to both when trying to use the clone current button


NOTE: configuration saving is **NOT** affected by this and a player can tab to someone's vehicle, then save it as a named configuration through the tuning menu, then spawn that vehicle. This is the same as cloning, but more steps, and we should see if we can prohibit this as well.

If I remember correctly, a difficulty here lies in a difference of access to different functions used by the clone current button and the configuration save button.

**Comment by lionkor** [comment]
> This is the same as cloning, but more steps, and we should see if we can prohibit this as well.

agreed, could we see to get that fixed also?

**Comment by Starystars67** [comment]
> configuration saving is **NOT** affected by this and a player can tab to someone's vehicle,

Looking over the code for how this works it calls partmgmt.saveLocal which is oveerwritten as part of this PR. 
See: https://github.com/BeamMP/BeamMP/pull/545/files#diff-0dea18c3fb9b9a9abe7baeef723ec9dbd98a148a5f3990947f084d1ab590fd42R998

Based on that this PR should address the case you mention. Do you have any suggestions as to why it does not?

**Comment by StanleyDudek** [comment]
> Based on that this PR should address the case you mention. Do you have any suggestions as to why it does not?

To the best of my understanding, the html/js side of the save menu UI loads and holds the original save function from the base game, that while should be overridden as the pr indicates, it is not. Curious to note that this replacement approach does work for the clone current button, but, comparing the html/js of the vehicle select menu (where clone current button is) vs vehicle config menu (where save/load is via partial.save.html) that clone current is called from vehiclselect.js with `bngApi.engineLua('core_vehicles.cloneCurrent()', ...` and save vehicle is called from vehicleconfig.js with `bngApi.engineLua(`extensions.core_vehicle_partmgmt.saveLocal("${configName}.pc")`)`

I'm not entirely sure beyond this what a solution looks like.

File: lua/ge/extensions/MPVehicleGE.lua [hunk 1]
```diff
@@ -928,7 +928,14 @@ local function sendVehicleSpawn(gameVehicleID)
 		vehicleTable.vcf = vehicleData.config -- Vehicle Config, contains paint data
 		vehicleTable.pos = {pos.x, pos.y, pos.z} -- Position
 		vehicleTable.rot = {rot.x, rot.y, rot.z, rot.w} -- Rotation
+		vehicleTable.pro = settings.getValue("protectConfigFromClone", false) -- Should the config be protected?
```
Comment (lionkor) [inline lua/ge/extensions/MPVehicleGE.lua:1068]: Could we name this properly so we don't need the comment? For example `protected`?

File: lua/ge/extensions/MPVehicleGE.lua [hunk 2]
```diff
@@ -928,7 +928,14 @@ local function sendVehicleSpawn(gameVehicleID)
...
+		if vehicleTable.pro == true then
+			vehicleTable.pro = "1"
+		else
+			vehicleTable.pro = "0"
+		end
```
Comment (lionkor) [inline lua/ge/extensions/MPVehicleGE.lua:1073]: Is there a specific reason why we don't use a bool here?

Comment (Starystars67) [reply lua/ge/extensions/MPVehicleGE.lua:1073]: The game gets funny with it sadly

File: lua/ge/extensions/MPVehicleGE.lua (hunk 1 above)
Comment (Starystars67) [reply lua/ge/extensions/MPVehicleGE.lua:1068]: I didnt want any issues to potentially arise in conflict with anything of the game or other details.

**Review by stefan750 — APPROVED** [review]

**Pull request merged**